### PR TITLE
[Natalia Falinska] lazy load

### PR DESCRIPTION
**Pull request by NFalinska**

### CODE REVIEW

**Comment by NFalinska** [comment]
https://issues.griddynamics.net/secure/RapidBoard.jspa?rapidView=298&view=detail&selectedIssue=UIT-40

File: src/app/product-list-page/components/product-list-page/product-list-page.component.html [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <div class="main-container">
-  <div class="filters"></div>
-  <div class="products-grid"></div>
-</div>
-<app-join-us></app-join-us>
+    <div class="filters"></div>
+    <div class="products-grid"></div>
+  </div>
```
Comment (Roman-Klimkevych) [inline src/app/product-list-page/components/product-list-page/product-list-page.component.html]: why do you need this change?

**Review by Roman-Klimkevych — APPROVED** [review]

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 2]
```diff
@@ -33,6 +33,7 @@ export class SlideshowComponent implements OnInit, AfterViewInit, OnDestroy {
     'assets/img/slideshow/banner-main2-1540x650.jpg'
   ];
 
+
```
Comment (Roman-Klimkevych) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: Do you need this extra blanc line here?

File: src/app/home-page/components/slideshow/slideshow.component.html [hunk 3]
```diff
@@ -3,7 +3,9 @@
     <div class="slides-holder" #slidesHolder>
       <div class="slide-item" [ngClass]="{animate: selectedSlideIndex === i}" #item [style.width.%]="translateStep"
         *ngFor="let img of responseImgs; let i = index">
-        <img [src]=img alt="slide">
+        <app-lazy-load>
+            <img [attr.data-src]="img" class="lazy-load" alt="slide">
+          </app-lazy-load>
```
Comment (Roman-Klimkevych) [inline src/app/home-page/components/slideshow/slideshow.component.html]: Indentation is a little bit off here.

File: src/app/home-page/components/slideshow/slideshow.component.spec.ts [hunk 4]
```diff
@@ -15,7 +16,7 @@ describe('SlideshowComponent', () => {
 
   beforeEach(async(() => {
     TestBed.configureTestingModule({
-      declarations: [ SlideshowComponent, ImgPlaceholderComponent ],
+      declarations: [ SlideshowComponent, ImgPlaceholderComponent, LazyLoadComponent],
       imports: [HttpClientTestingModule]
```
Comment (Roman-Klimkevych) [inline src/app/home-page/components/slideshow/slideshow.component.spec.ts:20]: You have to be consistent with the spaces in brackets.

File: src/app/product-details-page/components/product-description/product-description.component.spec.ts [hunk 5]
```diff
@@ -5,6 +5,7 @@ import { ZoomComponent } from './product-image-preview/zoom/zoom.component';
 import { ProductImagePreviewComponent } from './product-image-preview/product-image-preview.component';
 import { ProductOrderComponent } from './product-order/product-order.component';
 import { HttpClientTestingModule } from '@angular/common/http/testing';
+import { LazyLoadComponent } from 'src/app/shared/components/lazy-load/lazy-load.component';
```
Comment (Roman-Klimkevych) [inline src/app/product-details-page/components/product-description/product-description.component.spec.ts]: You import this component here but don't use it.

File: src/app/product-details-page/components/product-description/product-image-preview/product-image-preview.component.spec.ts [hunk 6]
```diff
@@ -4,14 +4,15 @@ import { RouterTestingModule } from '@angular/router/testing';
 import { ProductDescriptionComponent } from '../product-description.component';
 import { ProductOrderComponent } from '../product-order/product-order.component';
 import { ZoomComponent } from './zoom/zoom.component';
+import { LazyLoadComponent } from 'src/app/shared/components/lazy-load/lazy-load.component';
```
Comment (Roman-Klimkevych) [inline src/app/product-details-page/components/product-description/product-image-preview/product-image-preview.component.spec.ts]: You import this component here but don't use it.

File: src/app/shared/components/lazy-load/lazy-load.component.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  loadImage = (target) => {
+    const isBackground = !!target.dataset.bgSrc;
+    const src = isBackground ? target.dataset.bgSrc : target.dataset.src;
+    if (!src) { return; }
+
+    this.fetchImage(src).then(() => {
+      this.observer.unobserve(target);
+      target.classList.add('loadeding');
```
Comment (Roman-Klimkevych) [inline src/app/shared/components/lazy-load/lazy-load.component.ts]: a spelling error `'loadeding'`

File: src/app/product-details-page/components/product-description/product-image-preview/product-image-preview.component.spec.ts [hunk 8]
```diff
@@ -4,14 +4,15 @@ import { RouterTestingModule } from '@angular/router/testing';
...
 
 describe('ProductImagePreviewComponent', () => {
   let component: ProductImagePreviewComponent;
   let fixture: ComponentFixture<ProductImagePreviewComponent>;
 
   beforeEach(async(() => {
     TestBed.configureTestingModule({
-      declarations: [ProductImagePreviewComponent, ProductDescriptionComponent, ProductOrderComponent, ZoomComponent],
+      declarations: [ProductImagePreviewComponent, ProductDescriptionComponent, ProductOrderComponent, ZoomComponent ],
```
Comment (Roman-Klimkevych) [inline src/app/product-details-page/components/product-description/product-image-preview/product-image-preview.component.spec.ts]: When you have an array spreading over several lines, I think it's better to have each item on a separate line, like this:
```
declarations: [
  ProductImagePreviewComponent,
  ProductDescriptionComponent,
  ProductOrderComponent,
  ZoomComponent
],
```

File: src/app/shared/components/lazy-load/lazy-load.component.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
+import { Component, AfterViewInit, ElementRef } from '@angular/core';
+
+@Component({
+  selector: 'app-lazy-load',
+  templateUrl: './lazy-load.component.html',
+  styleUrls: ['./lazy-load.component.scss']
+})
+export class LazyLoadComponent implements AfterViewInit {
+
+  observer: IntersectionObserver;
```
Comment (dverbovyi) [inline src/app/shared/components/lazy-load/lazy-load.component.ts]: this should be private

File: src/app/shared/components/lazy-load/lazy-load.component.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+  constructor(public element: ElementRef) { }
+
+  options: any = {
```
Comment (dverbovyi) [inline src/app/shared/components/lazy-load/lazy-load.component.ts]: make private

File: src/app/shared/components/lazy-load/lazy-load.component.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    root: null,
+    rootMargin: '0px',
+    threshold: 0.1
+  };
+
+  ngAfterViewInit() {
+    const images = this.element.nativeElement.querySelectorAll('.lazy-load');
```
Comment (dverbovyi) [inline src/app/shared/components/lazy-load/lazy-load.component.ts:22]: `if(!images) return`

File: src/app/shared/components/lazy-load/lazy-load.component.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    this.observer = new IntersectionObserver(this.handleIntersection, this.options);
+    images.forEach(img => {
+      this.observer.observe(img);
+    });
+  }
+
+  fetchImage = (url) => {
```
Comment (dverbovyi) [inline src/app/shared/components/lazy-load/lazy-load.component.ts]: 1) make private
2) specify parameter type
3) it should be declared as class method not property: 
`private fetchImage(url: String):void { ... }`

File: src/app/shared/components/lazy-load/lazy-load.component.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    return new Promise((resolve, reject) => {
+      const image = new Image();
+      image.src = url;
+      image.onload = resolve;
+      image.onerror = reject;
+    });
+  }
+
+  loadImage = (target) => {
```
Comment (dverbovyi) [inline src/app/shared/components/lazy-load/lazy-load.component.ts]: this should be class method not arrow function
```ts
private loadImage(target: HTMLElement): void {...}
```

File: src/app/shared/components/lazy-load/lazy-load.component.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        target.style.backgroundImage = `url(${src})`;
+      } else {
+        target.removeAttribute('data-src');
+        target.src = src;
+      }
+      target.classList.add('loaded');
+    });
+  }
+
+  handleIntersection = (entries, observer) => {
```
Comment (dverbovyi) [inline src/app/shared/components/lazy-load/lazy-load.component.ts]: same here

File: src/app/shared/components/product-item/product-item-short/product-item-short.component.scss [hunk 15]
```diff
@@ -109,6 +109,15 @@ hr {
   z-index: 10;
 }
 
+.loading {
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item-short/product-item-short.component.scss]: you don't need this class

add default style directly for .lazy-load

File: src/app/shared/components/product-item/product-item.component.ts [hunk 16]
```diff
@@ -13,7 +13,7 @@ export class ProductItemComponent implements OnInit {
 
   isHovered = false;
   _currentThumbnail;
-  private _currentSwatch;
+  public _currentSwatch;
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item.component.ts]: it should be private

File: src/app/shared/components/carousel/carousel-item/carousel-item.component.spec.ts [hunk 17]
```diff
@@ -1,6 +1,7 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 
 import { CarouselItemComponent } from './carousel-item.component';
+import { LazyLoadComponent } from '../../lazy-load/lazy-load.component';
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel-item/carousel-item.component.spec.ts]: this component has not being used in this file

File: src/app/shared/components/carousel/carousel.component.spec.ts [hunk 18]
```diff
@@ -3,6 +3,7 @@ import { By } from '@angular/platform-browser';
 import { CarouselComponent } from './carousel.component';
 import { SharedModule } from '../../shared.module';
 import { ProductDetailsPageModule } from 'src/app/product-details-page/product-details-page.module';
+import { LazyLoadComponent } from '../lazy-load/lazy-load.component';
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.spec.ts]: this component has not being used in this file

File: src/app/shared/components/lazy-load/lazy-load.component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(LazyLoadComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
```
Comment (dverbovyi) [inline src/app/shared/components/lazy-load/lazy-load.component.spec.ts:22]: need more tests

File: src/app/shared/components/load-more/load-more.component.spec.ts [hunk 20]
```diff
@@ -1,6 +1,7 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 
 import { LoadMoreComponent } from './load-more.component';
+import { ProductItemComponent } from '../product-item/product-item.component';
```
Comment (dverbovyi) [inline src/app/shared/components/load-more/load-more.component.spec.ts]: this component has not being used in this file

File: src/app/shared/components/product-item/product-item.component.scss [hunk 21]
```diff
@@ -52,6 +53,11 @@
   }
 }
 
+.loaded {
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item.component.scss]: duplicated styles. Move to lazy-load.component.scss

File: src/app/shared/components/product-item/product-item.component.spec.ts [hunk 22]
```diff
@@ -60,6 +61,13 @@ describe('ProductItemComponent', () => {
     expect(component).toBeTruthy();
   });
 
+  it('should return right color', () => {
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item.component.spec.ts]: weird test. nothing in expect

**Review by dverbovyi — CHANGES_REQUESTED** [review]

File: src/app/shared/components/carousel/carousel.component.scss [hunk 23]
```diff
@@ -4,7 +4,7 @@
 @import "../../../../scss/mixins";
 
 .main-wrapper {
-  padding: 0 $content-padding;
+  margin: 0 $content-padding;
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.scss]: please revert this line

**Review by dverbovyi — APPROVED** [review]

File: src/app/shared/components/carousel/carousel.component.scss (hunk 23 above)
Comment (dverbovyi) [reply src/app/shared/components/carousel/carousel.component.scss]: @NFalinska

**Pull request merged**